File: sd/list/list.cpp
#include "list.hpp"
#include <iostream>
#include <string>

using namespace std;

List::List() : head(nullptr), tail(nullptr) {}

List::~List() {
    LNode* cur = head;
    while (cur) {
        LNode* tmp = cur;
        cur = cur->next;
        delete tmp;
    }
}

bool List::is_empty() const {
    return head == nullptr;
}

int List::get_size() const{
    int count = 0;
    LNode* cur = head;
    while(cur){
        count += 1;
        cur = cur->next;
    }
    return count;
}

List::LNode* List::find(const std::string& key) const {
    LNode* cur = head;
    while (cur) {
        if (cur->data == key) return cur;
        cur = cur->next;
    }
    return nullptr;
}

// Добавление в конец
void List::push_back(const string& val) {
    LNode* node = new LNode(val);
    if (is_empty()) {
        head = tail = node;
    } else {
        tail->next = node;
        tail = node;
    }
}

// Добавление в начало
void List::push_front(const string& val) {
    LNode* node = new LNode(val);
    node->next = head;
    head = node;
    if (!tail) tail = node;
}

// Вставка после заданного ключа
void List::insert_after(const string& key, const string& val) {
    LNode* cur = find(key);
    if (!cur) {
        cout << "Элемент '" << key << "' не найден.\n";
        return;
    }

    LNode* node = new LNode(val);
    node->next = cur->next;
    cur->next = node;
    if (cur == tail)
        tail = node;
}

// Вставка перед заданным ключом
void List::insert_before(const string& key, const string& val) {
    if (is_empty()) return;

    if (head->data == key) {
        push_front(val);
        return;
    }

    LNode* prev = nullptr;
    LNode* cur = head;

    while (cur && cur->data != key) {
        prev = cur;
        cur = cur->next;
    }

    if (!cur) {
        cout << "Элемент '" << key << "' не найден.\n";
        return;
    }

    LNode* node = new LNode(val);
    node->next = cur;
    prev->next = node;
}

// Удаление по значению
void List::del(const string& val) {
    if (is_empty()) return;

    if (head->data == val) {
        del_head();
        return;
    }

    LNode* prev = nullptr;
    LNode* cur = head;

    while (cur && cur->data != val) {
        prev = cur;
        cur = cur->next;
    }

    if (!cur) {
        cout << "Элемент '" << val << "' не найден.\n";
        return;
    }

    prev->next = cur->next;
    if (cur == tail)
        tail = prev;
    delete cur;
}

// Вывод списка
void List::print() const {
    LNode* cur = head;
    while (cur) {
        cout << cur->data << " ";
        cur = cur->next;
    }
    cout << endl;
}

// Получить элемент по индексу
List::LNode* List::get_at(int index) const {
    if (index < 0) return nullptr;
    LNode* cur = head;
    int i = 0;
    while (cur && i < index) {
        cur = cur->next;
        i++;
    }
    return cur;
}

// Удаление головы
void List::del_head() {
    if (is_empty()) return;

    LNode* tmp = head;
    head = head->next;
    delete tmp;

    if (!head) tail = nullptr;
}

// Удаление хвоста
void List::del_tail() {
    if (is_empty()) return;

    if (head == tail) {
        delete head;
        head = tail = nullptr;
        return;
    }

    LNode* cur = head;
    while (cur->next != tail)
        cur = cur->next;

    delete tail;
    tail = cur;
    tail->next = nullptr;
}

// Удалить элемент после ключа
void List::del_after(const string& key) {
    LNode* cur = find(key);
    if (!cur || !cur->next) return;

    LNode* tmp = cur->next;
    cur->next = tmp->next;
    if (tmp == tail)
        tail = cur;
    delete tmp;
}

// Удалить элемент перед ключом
void List::del_before(const string& key) {
    if (is_empty() || head->data == key) return;

    if (head->next && head->next->data == key) {
        del_head();
        return;
    }

    LNode* prevPrev = nullptr;
    LNode* prev = head;
    LNode* cur = head->next;

    while (cur && cur->data != key) {
        prevPrev = prev;
        prev = cur;
        cur = cur->next;
    }

    if (!cur) return; // не найден ключ

    prevPrev->next = cur;
    delete prev;
}

File: golang/pkg/datastructures/list.go
package datastructures

import (
	"bufio"
	"io"
	"strconv"
	"strings"
)

// Node представляет узел в односвязном списке
type Node struct {
	data string
	next *Node
}

// List - односвязный список
type List struct {
	head *Node
	tail *Node
	size int
}

// NewList создаёт новый пустой список
func NewList() *List {
	return &List{
		head: nil,
		tail: nil,
		size: 0,
	}
}

// IsEmpty проверяет, пуст ли список
func (l *List) IsEmpty() bool {
	return l.head == nil
}

// GetSize возвращает размер списка
func (l *List) GetSize() int {
	return l.size
}

// PushBack добавляет элемент в конец
func (l *List) PushBack(val string) {
	node := &Node{data: val}
	if l.IsEmpty() {
		l.head = node
		l.tail = node
	} else {
		l.tail.next = node
		l.tail = node
	}
	l.size++
}

// PushFront добавляет элемент в начало
func (l *List) PushFront(val string) {
	node := &Node{data: val}
	if l.IsEmpty() {
		l.head = node
		l.tail = node
	} else {
		node.next = l.head
		l.head = node
	}
	l.size++
}

// Find ищет элемент
func (l *List) Find(key string) *Node {
	current := l.head
	for current != nil {
		if current.data == key {
			return current
		}
		current = current.next
	}
	return nil
}

// GetAt получает элемент по индексу
func (l *List) GetAt(index int) *Node {
	if index < 0 || index >= l.size {
		return nil
	}
	current := l.head
	for i := 0; i < index; i++ {
		current = current.next
	}
	return current
}

// Del удаляет элемент по значению
func (l *List) Del(val string) error {
	if l.IsEmpty() {
		return ErrEmpty
	}

	if l.head.data == val {
		l.head = l.head.next
		if l.head == nil {
			l.tail = nil
		}
		l.size--
		return nil
	}

	current := l.head
	for current.next != nil {
		if current.next.data == val {
			current.next = current.next.next
			if current.next == nil {
				l.tail = current
			}
			l.size--
			return nil
		}
		current = current.next
	}
	return ErrNotFound
}

// Serialize сохраняет список в текстовом формате
func (l *List) Serialize(w io.Writer) error {
	// Сохраняем размер
	if _, err := w.Write([]byte(strconv.Itoa(l.size) + "\n")); err != nil {
		return err
	}

	// Сохраняем элементы
	current := l.head
	for current != nil {
		if _, err := w.Write([]byte(current.data + "\n")); err != nil {
			return err
		}
		current = current.next
	}
	return nil
}

// Deserialize загружает список из текстового формата
func (l *List) Deserialize(r io.Reader) error {
	// Очищаем текущий список
	l.head = nil
	l.tail = nil
	l.size = 0

	scanner := bufio.NewScanner(r)

	// Читаем размер
	if !scanner.Scan() {
		return scanner.Err()
	}
	sizeStr := strings.TrimSpace(scanner.Text())
	size, err := strconv.Atoi(sizeStr)
	if err != nil {
		return err
	}

	// Читаем элементы
	for i := 0; i < size; i++ {
		if !scanner.Scan() {
			return scanner.Err()
		}
		val := strings.TrimSpace(scanner.Text())
		l.PushBack(val)
	}

	return nil
}
